File: Capacitor-CORS2/ios/Plugin/CorsBypassPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(CorsBypassPlugin)
public class CorsBypassPlugin: CAPPlugin {
    private var sseConnections: [String: SSEConnection] = [:]
    private var connectionCounter = 0
    
    @objc func request(_ call: CAPPluginCall) {
        guard let url = call.getString("url") else {
            call.reject("URL is required")
            return
        }
        
        let method = call.getString("method") ?? "GET"
        let headers = call.getObject("headers") as? [String: String] ?? [:]
        let data = call.getValue("data")
        let params = call.getObject("params") as? [String: String]
        let timeout = call.getDouble("timeout") ?? 30.0
        let responseType = call.getString("responseType") ?? "json"
        let withCredentials = call.getBool("withCredentials") ?? false
        
        makeHttpRequest(
            url: url,
            method: method,
            headers: headers,
            data: data,
            params: params,
            timeout: timeout,
            responseType: responseType,
            withCredentials: withCredentials,
            call: call
        )
    }
    
    @objc func get(_ call: CAPPluginCall) {
        handleHttpMethod(call: call, method: "GET")
    }
    
    @objc func post(_ call: CAPPluginCall) {
        handleHttpMethod(call: call, method: "POST")
    }
    
    @objc func put(_ call: CAPPluginCall) {
        handleHttpMethod(call: call, method: "PUT")
    }
    
    @objc func patch(_ call: CAPPluginCall) {
        handleHttpMethod(call: call, method: "PATCH")
    }
    
    @objc func delete(_ call: CAPPluginCall) {
        handleHttpMethod(call: call, method: "DELETE")
    }
    
    private func handleHttpMethod(call: CAPPluginCall, method: String) {
        guard let url = call.getString("url") else {
            call.reject("URL is required")
            return
        }
        
        let headers = call.getObject("headers") as? [String: String] ?? [:]
        let data = call.getValue("data")
        let params = call.getObject("params") as? [String: String]
        let timeout = call.getDouble("timeout") ?? 30.0
        let responseType = call.getString("responseType") ?? "json"
        let withCredentials = call.getBool("withCredentials") ?? false
        
        makeHttpRequest(
            url: url,
            method: method,
            headers: headers,
            data: data,
            params: params,
            timeout: timeout,
            responseType: responseType,
            withCredentials: withCredentials,
            call: call
        )
    }
    
    private func makeHttpRequest(
        url: String,
        method: String,
        headers: [String: String],
        data: Any?,
        params: [String: String]?,
        timeout: Double,
        responseType: String,
        withCredentials: Bool,
        call: CAPPluginCall
    ) {
        var urlString = url
        
        // Add URL parameters
        if let params = params, !params.isEmpty {
            var urlComponents = URLComponents(string: url)
            var queryItems = urlComponents?.queryItems ?? []
            
            for (key, value) in params {
                queryItems.append(URLQueryItem(name: key, value: value))
            }
            
            urlComponents?.queryItems = queryItems
            urlString = urlComponents?.url?.absoluteString ?? url
        }
        
        guard let requestUrl = URL(string: urlString) else {
            call.reject("Invalid URL")
            return
        }
        
        var request = URLRequest(url: requestUrl)
        request.httpMethod = method
        request.timeoutInterval = timeout
        
        // Set headers
        for (key, value) in headers {
            request.setValue(value, forHTTPHeaderField: key)
        }
        
        // Set body for non-GET requests
        if method != "GET", let data = data {
            do {
                if let jsonData = data as? [String: Any] {
                    request.httpBody = try JSONSerialization.data(withJSONObject: jsonData)
                    if request.value(forHTTPHeaderField: "Content-Type") == nil {
                        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    }
                } else if let stringData = data as? String {
                    request.httpBody = stringData.data(using: .utf8)
                }
            } catch {
                call.reject("Failed to serialize request body: \(error.localizedDescription)")
                return
            }
        }
        
        // Configure session
        let config = URLSessionConfiguration.default
        if withCredentials {
            config.httpCookieAcceptPolicy = .always
            config.httpShouldSetCookies = true
        }
        
        let session = URLSession(configuration: config)
        
        let task = session.dataTask(with: request) { data, response, error in
            DispatchQueue.main.async {
                if let error = error {
                    call.reject("Request failed: \(error.localizedDescription)")
                    return
                }
                
                guard let httpResponse = response as? HTTPURLResponse else {
                    call.reject("Invalid response")
                    return
                }
                
                // Convert headers to dictionary
                var responseHeaders: [String: String] = [:]
                for (key, value) in httpResponse.allHeaderFields {
                    if let keyString = key as? String, let valueString = value as? String {
                        responseHeaders[keyString] = valueString
                    }
                }
                
                // Parse response data
                var responseData: Any = ""
                if let data = data {
                    switch responseType {
                    case "text":
                        responseData = String(data: data, encoding: .utf8) ?? ""
                    case "json":
                        do {
                            responseData = try JSONSerialization.jsonObject(with: data, options: [])
                        } catch {
                            responseData = String(data: data, encoding: .utf8) ?? ""
                        }
                    case "blob", "arraybuffer":
                        responseData = data.base64EncodedString()
                    default:
                        do {
                            responseData = try JSONSerialization.jsonObject(with: data, options: [])
                        } catch {
                            responseData = String(data: data, encoding: .utf8) ?? ""
                        }
                    }
                }
                
                call.resolve([
                    "data": responseData,
                    "status": httpResponse.statusCode,
                    "statusText": HTTPURLResponse.localizedString(forStatusCode: httpResponse.statusCode),
                    "headers": responseHeaders,
                    "url": httpResponse.url?.absoluteString ?? url
                ])
            }
        }
        
        task.resume()
    }
    
    @objc func startSSE(_ call: CAPPluginCall) {
        guard let url = call.getString("url") else {
            call.reject("URL is required")
            return
        }
        
        connectionCounter += 1
        let connectionId = "sse_\(connectionCounter)"
        
        let headers = call.getObject("headers") as? [String: String] ?? [:]
        let withCredentials = call.getBool("withCredentials") ?? false
        let reconnectTimeout = call.getDouble("reconnectTimeout") ?? 3.0
        
        let sseConnection = SSEConnection(
            url: url,
            headers: headers,
            withCredentials: withCredentials,
            reconnectTimeout: reconnectTimeout,
            plugin: self,
            connectionId: connectionId
        )
        
        sseConnections[connectionId] = sseConnection
        sseConnection.connect()
        
        call.resolve(["connectionId": connectionId])
    }
    
    @objc func stopSSE(_ call: CAPPluginCall) {
        guard let connectionId = call.getString("connectionId") else {
            call.reject("Connection ID is required")
            return
        }
        
        if let connection = sseConnections[connectionId] {
            connection.disconnect()
            sseConnections.removeValue(forKey: connectionId)
        }
        
        call.resolve()
    }
    
    // SSE event handlers
    func notifySSEOpen(connectionId: String) {
        notifyListeners("sseOpen", data: [
            "connectionId": connectionId,
            "status": "connected"
        ])
    }
    
    func notifySSEMessage(connectionId: String, data: String, id: String?, type: String?) {
        var eventData: [String: Any] = [
            "connectionId": connectionId,
            "data": data
        ]
        
        if let id = id {
            eventData["id"] = id
        }
        
        if let type = type {
            eventData["type"] = type
        }
        
        notifyListeners("sseMessage", data: eventData)
    }
    
    func notifySSEError(connectionId: String, error: String) {
        notifyListeners("sseError", data: [
            "connectionId": connectionId,
            "error": error
        ])
    }
    
    func notifySSEClose(connectionId: String) {
        notifyListeners("sseClose", data: [
            "connectionId": connectionId,
            "status": "disconnected"
        ])
    }
}

File: Capacitor-CORS2/android/src/main/java/com/capacitor/cors/CorsBypassPlugin.java
package com.capacitor.cors;

import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.io.IOException;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.concurrent.TimeUnit;
import okhttp3.*;
import org.json.JSONException;
import org.json.JSONObject;

@CapacitorPlugin(name = "CorsBypass")
public class CorsBypassPlugin extends Plugin {
    private static final String TAG = "CorsBypassPlugin";
    private final Map<String, SSEConnection> sseConnections = new HashMap<>();
    private int connectionCounter = 0;
    private OkHttpClient httpClient;

    @Override
    public void load() {
        super.load();
        
        // Initialize HTTP client
        httpClient = new OkHttpClient.Builder()
            .connectTimeout(30, TimeUnit.SECONDS)
            .readTimeout(30, TimeUnit.SECONDS)
            .writeTimeout(30, TimeUnit.SECONDS)
            .build();
    }

    @PluginMethod
    public void request(PluginCall call) {
        makeHttpRequest(call, call.getString("method", "GET"));
    }

    @PluginMethod
    public void get(PluginCall call) {
        makeHttpRequest(call, "GET");
    }

    @PluginMethod
    public void post(PluginCall call) {
        makeHttpRequest(call, "POST");
    }

    @PluginMethod
    public void put(PluginCall call) {
        makeHttpRequest(call, "PUT");
    }

    @PluginMethod
    public void patch(PluginCall call) {
        makeHttpRequest(call, "PATCH");
    }

    @PluginMethod
    public void delete(PluginCall call) {
        makeHttpRequest(call, "DELETE");
    }

    private void makeHttpRequest(PluginCall call, String method) {
        String url = call.getString("url");
        if (url == null) {
            call.reject("URL is required");
            return;
        }

        JSObject headers = call.getObject("headers", new JSObject());
        JSObject params = call.getObject("params", new JSObject());
        Object data = call.getData().opt("data");
        double timeout = call.getDouble("timeout", 30.0);
        String responseType = call.getString("responseType", "json");
        boolean withCredentials = call.getBoolean("withCredentials", false);

        try {
            // Build URL with parameters
            HttpUrl.Builder urlBuilder = HttpUrl.parse(url).newBuilder();
            Iterator<String> paramKeys = params.keys();
            while (paramKeys.hasNext()) {
                String key = paramKeys.next();
                urlBuilder.addQueryParameter(key, params.getString(key));
            }
            HttpUrl requestUrl = urlBuilder.build();

            // Build request
            Request.Builder requestBuilder = new Request.Builder().url(requestUrl);

            // Add headers
            Iterator<String> headerKeys = headers.keys();
            while (headerKeys.hasNext()) {
                String key = headerKeys.next();
                requestBuilder.addHeader(key, headers.getString(key));
            }

            // Add body for non-GET requests
            RequestBody body = null;
            if (!method.equals("GET") && data != null) {
                MediaType mediaType = MediaType.parse("application/json; charset=utf-8");
                if (data instanceof JSONObject) {
                    body = RequestBody.create(mediaType, data.toString());
                } else if (data instanceof String) {
                    body = RequestBody.create(mediaType, (String) data);
                }
                
                if (!headers.has("Content-Type")) {
                    requestBuilder.addHeader("Content-Type", "application/json");
                }
            }

            requestBuilder.method(method, body);
            Request request = requestBuilder.build();

            // Configure client with timeout
            OkHttpClient client = httpClient.newBuilder()
                .connectTimeout((long) timeout, TimeUnit.SECONDS)
                .readTimeout((long) timeout, TimeUnit.SECONDS)
                .writeTimeout((long) timeout, TimeUnit.SECONDS)
                .build();

            // Execute request
            client.newCall(request).enqueue(new Callback() {
                @Override
                public void onFailure(Call httpCall, IOException e) {
                    call.reject("Request failed: " + e.getMessage());
                }

                @Override
                public void onResponse(Call httpCall, Response response) throws IOException {
                    try {
                        // Parse response headers
                        JSObject responseHeaders = new JSObject();
                        for (String name : response.headers().names()) {
                            responseHeaders.put(name, response.header(name));
                        }

                        // Parse response body
                        Object responseData = "";
                        if (response.body() != null) {
                            String bodyString = response.body().string();
                            
                            switch (responseType) {
                                case "text":
                                    responseData = bodyString;
                                    break;
                                case "json":
                                    try {
                                        responseData = new JSONObject(bodyString);
                                    } catch (JSONException e) {
                                        responseData = bodyString;
                                    }
                                    break;
                                case "blob":
                                case "arraybuffer":
                                    // Convert to base64 for transfer
                                    responseData = android.util.Base64.encodeToString(
                                        bodyString.getBytes(), android.util.Base64.DEFAULT);
                                    break;
                                default:
                                    try {
                                        responseData = new JSONObject(bodyString);
                                    } catch (JSONException e) {
                                        responseData = bodyString;
                                    }
                                    break;
                            }
                        }

                        JSObject result = new JSObject();
                        result.put("data", responseData);
                        result.put("status", response.code());
                        result.put("statusText", response.message());
                        result.put("headers", responseHeaders);
                        result.put("url", response.request().url().toString());

                        call.resolve(result);
                    } catch (Exception e) {
                        call.reject("Failed to parse response: " + e.getMessage());
                    }
                }
            });

        } catch (Exception e) {
            call.reject("Failed to create request: " + e.getMessage());
        }
    }

    @PluginMethod
    public void startSSE(PluginCall call) {
        String url = call.getString("url");
        if (url == null) {
            call.reject("URL is required");
            return;
        }

        connectionCounter++;
        String connectionId = "sse_" + connectionCounter;

        JSObject headers = call.getObject("headers", new JSObject());
        boolean withCredentials = call.getBoolean("withCredentials", false);
        double reconnectTimeout = call.getDouble("reconnectTimeout", 3.0);

        try {
            SSEConnection sseConnection = new SSEConnection(
                url, headers, withCredentials, reconnectTimeout, this, connectionId, httpClient
            );
            
            sseConnections.put(connectionId, sseConnection);
            sseConnection.connect();

            JSObject result = new JSObject();
            result.put("connectionId", connectionId);
            call.resolve(result);

        } catch (Exception e) {
            call.reject("Failed to start SSE connection: " + e.getMessage());
        }
    }

    @PluginMethod
    public void stopSSE(PluginCall call) {
        String connectionId = call.getString("connectionId");
        if (connectionId == null) {
            call.reject("Connection ID is required");
            return;
        }

        SSEConnection connection = sseConnections.get(connectionId);
        if (connection != null) {
            connection.disconnect();
            sseConnections.remove(connectionId);
        }

        call.resolve();
    }

    // SSE event handlers
    public void notifySSEOpen(String connectionId) {
        JSObject data = new JSObject();
        data.put("connectionId", connectionId);
        data.put("status", "connected");
        notifyListeners("sseOpen", data);
    }

    public void notifySSEMessage(String connectionId, String messageData, String id, String type) {
        JSObject data = new JSObject();
        data.put("connectionId", connectionId);
        data.put("data", messageData);
        
        if (id != null) {
            data.put("id", id);
        }
        
        if (type != null) {
            data.put("type", type);
        }
        
        notifyListeners("sseMessage", data);
    }

    public void notifySSEError(String connectionId, String error) {
        JSObject data = new JSObject();
        data.put("connectionId", connectionId);
        data.put("error", error);
        notifyListeners("sseError", data);
    }

    public void notifySSEClose(String connectionId) {
        JSObject data = new JSObject();
        data.put("connectionId", connectionId);
        data.put("status", "disconnected");
        notifyListeners("sseClose", data);
    }
}
